ThreeTen/threeten-extra
# Upgrade to JUnit 5

Upgrades to JUnit 5, per #172.

I figure as I've done several JUnit 5 migrations in my day job that this would be a straightforward way to contribute back to this project.

## JUnit dataprovider

There were [multiple migration paths suggested](https://github.com/TNG/junit-dataprovider/wiki/Migration-guides#migration-to-junit5) for the [JUnit dataprovider](https://github.com/TNG/junit-dataprovider) library.  I went with the one that utilizes the [JUnit 5 Parameterized Tests feature](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests).  I followed the provided [migration guide](https://github.com/TNG/junit-dataprovider/wiki/Migration-from-JUnit4-to-JUnit-Jupiter-Parameterized-Tests).

Aside from the library version change and imports, this involved changing all `@UseDataProvider` tests to be annotated with `@ParameterizedTest` rather than `@Test`, and removing `@RunWith(DataProviderRunner.class)`.

## JUnit 5 changes

Aside from imports, the bulk of the changes changes was JUnit 5's removal of [`expected`](https://junit.org/junit4/javadoc/latest/org/junit/Test.html#expected()) from the `@Test` annotation in favor of [`assertThrows`](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertThrows(java.lang.Class,org.junit.jupiter.api.function.Executable)).

This had the benefit of improving some tests to more precisely declare _where_ the exception is to be thrown.  Additionally, some existing weren't fully executing, since the first call that threw the expected exception stopped the running of the test.  For example, the second `eraOf` call wouldn't have run in the following:

```diff
-    @Test(expected = DateTimeException.class)
+    @Test
     public void test_Chronology_eraOf_invalid() {
-        DiscordianChronology.INSTANCE.eraOf(2);
-        DiscordianChronology.INSTANCE.eraOf(0);
+        assertThrows(DateTimeException.class, () -> DiscordianChronology.INSTANCE.eraOf(2));
+        assertThrows(DateTimeException.class, () -> DiscordianChronology.INSTANCE.eraOf(0));
     }
```

I also took the extra step of changing any location in the code that used `fail()` within a try/catch block to instead use `assertThrows`.

```diff
-                try {
-                    sample.range(field);
-                    fail("Failed on " + sample + " " + field);
-                } catch (DateTimeException ex) {
-                    // expected
-                }
+                assertThrows(DateTimeException.class, () -> sample.range(field), "Failed on " + sample + " " + field);
```

## Maven dependency versions

This is now declaring the JUnit version three times in the pom file:

```xml
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <version>5.7.1</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-params</artifactId>
      <version>5.7.1</version>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <version>5.7.1</version>
      <scope>test</scope>
    </dependency>
```

I wasn't sure how you would want the duplication handled, so I left it duplicated for now.  The three options I see are:
1. Keep it duplicated
2. Use a [Maven property](https://maven.apache.org/pom.html#Properties)
3. Use the [JUnit BOM](https://junit.org/junit5/docs/current/user-guide/#dependency-metadata-junit-bom)
   ```xml
     <dependency>
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter-api</artifactId>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter-params</artifactId>
       <scope>test</scope>
     </dependency>
     <dependency>
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter-engine</artifactId>
       <scope>test</scope>
     </dependency>
   </dependencies>
 
   <dependencyManagement>
     <dependencies>
       <dependency>
         <groupId>org.junit</groupId>
         <artifactId>junit-bom</artifactId>
         <version>5.7.1</version>
         <type>pom</type>
         <scope>import</scope>
       </dependency>
     </dependencies>
   </dependencyManagement>
   ```
